### PR TITLE
fix: allow to migrate files that already have a .cy.

### PR DESCRIPTION
**Pull request by elevatebart**
- Closes UNIFY-1071

### User facing changelog

We can now migrate the realworld-app files twice.

### How has the user experience changed?

Before
Screen Shot 2022-02-17 at 2 51 13 PM

After
Screen Shot 2022-02-17 at 2 48 38 PM

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/test/unit/sources/migration/autoRename.spec.ts [hunk 1]
```diff
@@ -393,6 +393,60 @@ describe('applyMigrationTransform', () => {
       expect(result.before).to.eql(expected.before)
       expect(result.after).to.eql(expected.after)
     })
+
```
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/migration/autoRename.spec.ts:396]: Nice. Sorry to slow you down but can we add an additional integration test in `packages/launchpad/cypress/e2e/migration.cy.ts` to verify this works? Ideally we'd add a new project to `system-tests`, maybe called `migration-already-migrated` that showcases this workflow.

What do you think? Too heavy handed? Main concern is around quality here, I think this unit test covers the use case, but I want to be sure (eg, it doesn't verify what happens when you hit rename, or skip, etc, for an already migrated project).

**Comment by lmiller1990** [comment]
@elevatebart can you add someone from E2E to review this?

File: packages/data-context/test/unit/sources/migration/autoRename.spec.ts (hunk 1 above)
Comment (elevatebart) [reply packages/data-context/test/unit/sources/migration/autoRename.spec.ts:396]: great idea !

Comment (elevatebart) [reply packages/data-context/test/unit/sources/migration/autoRename.spec.ts:396]: done

Comment (lmiller1990) [reply packages/data-context/test/unit/sources/migration/autoRename.spec.ts:396]: Can you update the system test to actually click through the entire migration flow? Also, don't forget they can try to skip renaming, etc. Currently it is basically just a UI test.

Comment (elevatebart) [reply packages/data-context/test/unit/sources/migration/autoRename.spec.ts:396]: I updated the rest of the test, though its value seems to be lower.

**Review by lmiller1990 — APPROVED** [review]

**Review by lmiller1990 — DISMISSED** [review]

**Comment by emilyrohrbough** [comment]
I'm confused with the specs being listed in the before & after with the same names. If no specs were moved, shouldn't we show a message saying, "No specs were moved" or something along those lines? If you had quite a few specs, but nothing changes, you'd end up doing a long comparison to see what truly changes to determine nothing did.

**Comment by elevatebart** [comment]
@emilyrohrbough I agree with your assessment. Showing files that will not be renamed is confusing and weird.
I am going to attempt to filter them out.

From what I remember, it had unforeseen consequences.

Regarding the text change, this it is a rare case IMO. 
They could also use a similar `.cy.` pattern in their file naming prior to migrating to avoid conflict.
We do not know.

I will see how much more time I can spend on that seemingly rare issue.

**Review by lmiller1990 — DISMISSED** [review]

**Comment by brian-mann** [comment]
I agree with @emilyrohrbough - the intention of the UI was to only show files that are being migrated - if there is no change from "before / after" those files should obviously be filtered out. If there are no files to migrate, then Step 1 is skipped entirely.

File: packages/data-context/src/sources/MigrationDataSource.ts [hunk 2]
```diff
@@ -197,13 +197,13 @@ export class MigrationDataSource {
...
 
-    const canBeAutomaticallyMigrated: MigrationFile[] = specs.integration.map(applyMigrationTransform)
+    const canBeAutomaticallyMigrated: MigrationFile[] = specs.integration.map(applyMigrationTransform).filter((spec) => spec.before.relative !== spec.after.relative)
 
     const defaultComponentPattern = isDefaultTestFiles(await this.parseCypressConfig(), 'component')
 
     // Can only migration component specs if they use the default testFiles pattern.
     if (defaultComponentPattern) {
-      canBeAutomaticallyMigrated.push(...specs.component.map(applyMigrationTransform))
+      canBeAutomaticallyMigrated.push(...specs.component.map(applyMigrationTransform).filter((spec) => spec.before.relative !== spec.after.relative))
```
Comment (emilyrohrbough) [inline packages/data-context/src/sources/MigrationDataSource.ts:206]:
```suggestion
      canBeAutomaticallyMigrated.push(...canBeAutomaticallyMigrated)
```

File: packages/launchpad/cypress/e2e/migration.cy.ts [hunk 3]
```diff
@@ -535,12 +535,7 @@ describe('Full migration flow for each project', { retries: { openMode: 2, runMo
 
   it('handles re-migrating a partially migrated codebase', { retries: 0 }, () => {
```
Comment (emilyrohrbough) [inline packages/launchpad/cypress/e2e/migration.cy.ts:536]: Can we update this test name to indicate it's a fully-migrated project? Do we need to dd a test for the partially migrated project as well?

File: packages/data-context/src/sources/MigrationDataSource.ts (hunk 2 above)
Comment (elevatebart) [reply packages/data-context/src/sources/MigrationDataSource.ts:206]: I do not understand what you mean.
Do you want to add the integration files twice in this array?

File: packages/launchpad/cypress/e2e/migration.cy.ts (hunk 3 above)
Comment (elevatebart) [reply packages/launchpad/cypress/e2e/migration.cy.ts:536]: This is not a fully migrated project yet. It still has a `cypress.json` file.

The scenario I am testing for here is someone has gone through the first step (autoRename) of the migration and closed cypress. When they re-open it it should not crash or break their setup.

Comment (emilyrohrbough) [reply packages/launchpad/cypress/e2e/migration.cy.ts:552]: Ahh gottchya. 👍🏻 The project name here wasn't super obvious to me that it was partially migrated.

File: packages/data-context/src/sources/MigrationDataSource.ts (hunk 2 above)
Comment (emilyrohrbough) [reply packages/data-context/src/sources/MigrationDataSource.ts:206]: Oh whoops, I thought you were mapping these values twice from specs.integration. I totally missed spec.**component**.

File: system-tests/projects/migration-already-migrated/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+## Migration E2E Defaults
+
+An e2e project with all defaults. We rename the `integrationFolder` and spec extension.
+
+The following migration steps will be used during this migration:
+
+- [x] automatic file rename
```
Comment (emilyrohrbough) [inline system-tests/projects/migration-already-migrated/README.md:7]:
```suggestion
- [] automatic file rename
```

File: system-tests/projects/migration-already-migrated/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+- [x] update config file
+- [ ] setup component testing
+
+## Automatic Migration
+
+Unless the user skips this step, after this step, the filesystem will be:
+
+| Before | After|
+|---|---|
+| `integration/basic.test.js` | `e2e/basic.cy.js` |
```
Comment (emilyrohrbough) [inline system-tests/projects/migration-already-migrated/README.md:19]: This step will be skipped

File: system-tests/projects/migration-already-migrated/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+## Migration E2E Defaults
+
+An e2e project with all defaults. We rename the `integrationFolder` and spec extension.
```
Comment (emilyrohrbough) [inline system-tests/projects/migration-already-migrated/README.md:3]: Lets rename this to something like `migration-specs-already-migrated`

**Review by emilyrohrbough — APPROVED** [review]

**Pull request merged**